google-cloud-storage/src/main/java/com/google/cloud/storage/MultipartUploadClientImpl.java

```diff
@@ -0,0 +1,153 @@
+
+  private final MultipartUploadHttpRequestManager httpRequestManager;
+  private final XmlMapper xmlMapper;
+  private final HttpStorageOptions options;
+  private final Retrier retrier;
+  private final URI uri;
+
+  public MultipartUploadClientImpl(
+      URI uri, HttpRequestFactory requestFactory, Retrier retrier, HttpStorageOptions options) {
+    this.httpRequestManager = new MultipartUploadHttpRequestManager(requestFactory);
+    this.xmlMapper = new XmlMapper();
+    this.options = options;
+    this.retrier = retrier;
+    this.uri = uri;
+  }
+
+  private Map<String, String> getGenericExtensionHeader() {
+    Map<String, String> extensionHeaders = new HashMap<>();
+    if (options.getClientLibToken() != null) {
+      extensionHeaders.put("x-goog-api-client", options.getClientLibToken());
+    }
+    if (options.getProjectId() != null) {
+      extensionHeaders.put("x-goog-user-project", options.getProjectId());
+    }
+    extensionHeaders.put("Date", getRfc1123Date());
+    return extensionHeaders;
+  }
+
+  public CreateMultipartUploadResponse createMultipartUpload(CreateMultipartUploadRequest request)
+      throws IOException {
+    String encodedBucket = encode(request.bucket());
+    String encodedKey = encode(request.key());
+    String resourcePath = "/" + encodedBucket + "/" + encodedKey;
+    String createUri = uri.toString() + resourcePath + "?uploads";
+
+    String contentType;
+    if (request.getContentType() == null) {
+      contentType = "application/x-www-form-urlencoded";
+    } else {
+      try {
+        contentType = MediaType.parse(request.getContentType()).toString();
+      } catch (IllegalArgumentException e) {
+        throw new IllegalArgumentException(
+            "Invalid Content-Type header provided: " + request.getContentType(), e);
+      }
+    }
+
+    HttpResponse response =
+        httpRequestManager.sendCreateMultipartUploadRequest(
+            createUri, contentType, request, getExtensionHeadersForCreateMultipartUpload(request));
+
+    if (!response.isSuccessStatusCode()) {
+      String error = response.parseAsString();
+      throw new RuntimeException(
+          "Failed to initiate upload: " + response.getStatusCode() + " " + error);
+    }
+
+    return xmlMapper.readValue(response.getContent(), CreateMultipartUploadResponse.class);
+  }
+
+  private Map<String, String> getExtensionHeadersForCreateMultipartUpload(
+      CreateMultipartUploadRequest request) {
+    Map<String, String> extensionHeaders = getGenericExtensionHeader();
+    if (request.getCannedAcl() != null) {
+      extensionHeaders.put("x-goog-acl", request.getCannedAcl().toString());
+    }
+    if (request.getMetadata() != null) {
+      for (Map.Entry<String, String> entry : request.getMetadata().entrySet()) {
+        if (entry.getKey() != null || entry.getValue() != null) {
+          extensionHeaders.put("x-goog-meta-" + entry.getKey(), entry.getValue());
```

Comment:
http header values should be URL encoded to ensure any reserved characters are not interpreted/impacting to request parsing.

Since the customer is able to provide arbitrary values here we need to encode them to make them http safe.

We should check with someone on GCS server side to see if they can handle encoded header keys as well.